### PR TITLE
Adds a new robotics category, Augmentations, along with 1 Augmentation.

### PR DESCRIPTION
**Pull request by Zonespace27**
[Balance] [Feature]
## About the PR
**NOTE**: This PR is #5828, but only adds the Augmentation framework and the Wireless Interactor. If this PR goes through, more Augmentations will likely be PR'ed in the future.

This PR adds a new type of organ for robotics to play around with, Augmentations.
Fundamentally, Augmentations are meant to be a risk/reward type system, with them giving an upside, along with some form of a downside.
They can be installed in a person by two different methods: First is by accessing the organ (in the case of brain augmentations, like the one here, it's by performing scalpel + saw on the head), then installing the augmentation by hitting the person with it (self surgery cannot install augmentations). The second is fully removing the organ, then attaching the augmentation to the organ itself, then reinstalling it. In both cases, you can remove it with any `TOOL_SNIPPING`. Cyberorgans and synth-organs cannot have augmentations installed.

Augmentation Added:
- Wireless Interactor: The user gains the ability to interact with objects at range, akin to cyborgs. However, there are some caveats with this. They can only interact with objects, so no remotely attacking mobs, and items cannot be touched, so remotely picking up items won't work either. The downside of this augmentation is that, like borgs, The user gains a form of flash vulnerability. When flashed, you take 25 stamina damage regardless of if you're flash-proof or not, along with to any additional stamina damage/disorient from the flash itself. In addition, your ranged interaction abilities are disabled for 15 seconds after being flashed, and finally, the augment takes 10 damage each time it's flashed. On breaking, the ranged touch is disabled, the flash stamina damage is upped to 40, and nanomachines are leaked into the user at 0.5u / life cycle.
Can be produced at a robotics fabricator, cost is: 8 Cry-1, 5 Met-2, 5 Con-2, 3 Ins-1, 10 telecrystal.

## Why's this needed? 
It'd be nice for robotics to have more voluntary organs they can shove into people, possibly giving them some business once in a while. In addition, it provides yet another sink for materials, requiring cooperation from mining. Also, it gives the player some agency in how they want to modify their character, if at all, unlike regular (cyber)organs. Finally, once augmentations to body organs are added (framework for them is mostly in place), they will be an alternative to making the organ a cybernetic one.

## Changelog

```changelog
(u)Zonespace
(*)Added a new kind of robotics modification, found at robotics fabricators, Augmentations! More details in minor changes.
(+)Augmentations are a "risk/reward" kind of deal, with a tradeoff.
(+)Wireless Interaction Augmentation: You can interact with objects at range, at the cost of flash vulnerability.
```

### CODE REVIEW

**Comment by Zonespace27** [comment]
The augment's sprite is a half stand-in, it will likely be improved upon in the future.

File: code/WorkInProgress/Electronics.dm [hunk 1]
```diff
@@ -783,7 +783,7 @@
 /obj/machinery/rkit/Topic(href, href_list)
 	if (usr.stat)
 		return
-	if ((in_interact_range(src, usr) && istype(src.loc, /turf)) || (issilicon(usr)))
+	if ((in_interact_range(src, usr) && istype(src.loc, /turf)) || (issilicon(usr)) || HasWirelessInteractor(usr))
```
Comment (Azrun) [inline code/WorkInProgress/Electronics.dm:786]: Isn't this already covered by `in_interact_range`?  I guess this covers instances when the player is inside something (closet/pod)... 🤷

File: code/modules/vending/vending.dm [hunk 2]
```diff
@@ -633,11 +633,17 @@
 					var/atom/movable/vended = new product_path(src.get_output_location()) // changed from obj, because it could be a mob, THANKS VALUCHIMP
 					vended.layer = src.layer + 0.1 //So things stop spawning under the fukin thing
 					if(isitem(vended))
-						usr.put_in_hand_or_eject(vended) // try to eject it into the users hand, if we can
+						if(!get_dist(usr, src) > 1)
```
Comment (Azrun) [inline code/modules/vending/vending.dm:636]: Does that work? Isn't that always true?
(`!` has precedence over `>`)

File: code/WorkInProgress/Electronics.dm (hunk 1 above)
Comment (Zonespace27) [reply code/WorkInProgress/Electronics.dm:786]: I was trying to fix an issue, this fix didn't work, forgot to uncommit
fuck

**Comment by mordent-goonstation** [comment]
General opinion: I do not like this being treated as an organ, and would much prefer it be implemented as an alternative organ or something you can apply *to* an organ and then reinstall.

**Comment by Azrun** [comment]
> General opinion: I do not like this being treated as an organ, and would much prefer it be implemented as an alternative organ or something you can apply *to* an organ and then reinstall.


This! I originally imagined these as things you would shove into/onto someone's brain.  So works require assistance to perform.

File: code/obj/item/organs/augmentations.dm [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+			src.donor.remove_stamina(25)
+			SPAWN_DBG(15 SECONDS)
+				src.flashed = FALSE
+		if(src.broken)
+			src.donor.remove_stamina(15)
+
+	on_transplant(var/mob/M as mob)
+		..()
+		if(!broken)
+			RegisterSignal(src.donor, COMSIG_CLICK, .proc/ranged_click)
```
[the remark this answers is not in the document]
Comment (Zonespace27) [reply code/obj/item/organs/augmentations.dm:53]: I'm treating it like how silicon code does it, currently.

File: code/modules/vending/vending.dm (hunk 2 above)
Comment (Zonespace27) [reply code/modules/vending/vending.dm]: Fixed all issues you raised

**Comment by Zonespace27** [comment]
After taking a short break from coding, I have been back at recoding this, per Mordent's request. It should (probably) be done by the end of the week or so!

**Comment by Zonespace27** [comment]
@mordent-goonstation As of writing this, I believe I've fully re-coded the augmentation system to work WITH organs, instead of them being their own organs. I probably missed something, I'm tired and it's nearly 10 PM, so I'll take another poke at it tomorrow to make sure everything works and is cleanly coded.

All that's left is getting a couple sprites, but that doesn't affect the main code part.

**Comment by Zonespace27** [comment]
> Big thing from brief playtesting: this appears to end up calling `attack_hand` on things in addition to `attack_ai`, letting you pull out light tubes, bash down doors, etc from a distance. Which is a problem

fixed

**Review by Tarmunora — CHANGES_REQUESTED** [review]
Augmented humans are now incapable of performing item specials or firing guns.

**Comment by Zonespace27** [comment]
> Augmented humans are now incapable of performing item specials or firing guns.

(mostly) fixed, you can't fire guns or such if you're clicking on something that's machinery, unsure how to fix

**Comment by Zonespace27** [comment]
@Tarmunora fixed

File: code/obj/machinery.dm [hunk 4]
```diff
@@ -134,22 +134,25 @@
...
+		H = usr
 	if(status & (NOPOWER|BROKEN))
 		//boutput(usr, "<span class='alert'>That machine is not powered!</span>")
 		return 1
 	if(usr.restrained() || usr.lying || usr.stat)
 		//boutput(usr, "<span class='alert'>You are unable to do that currently!</span>")
 		return 1
 	if(!hasvar(src,"portable") || !src:portable)
-		if ((!in_interact_range(src, usr) || !istype(src.loc, /turf)) && !issilicon(usr) && !isAI(usr))
+		if ((!in_interact_range(src, usr) || !istype(src.loc, /turf)) && !issilicon(usr) && !isAI(usr) && !locate(/obj/item/augmentation/head/wireless_interact) in H?.has_augmentation("Head"))
```
Comment (ZeWaka) [inline code/obj/machinery.dm:147]: this syntax is really weird? locate a typepath in a has_augmentation call return?

Seems really clunky to me, should be wrapped into a call

File: code/obj/item/organ_holder.dm [hunk 5]
```diff
@@ -713,6 +732,225 @@
...
+
+			if ("right_eye")
+				if (!src.right_eye)
+					return 0
+				var/obj/item/augmentation/myAug = src.right_eye?.installed_aug
+				src.right_eye.installed_aug.set_loc(location)
+				src.right_eye.installed_aug.on_organ_removal()
+				src.right_eye.installed_aug = null
+				src.donor.update_body()
+				return myAug
```
Comment (ZeWaka) [inline code/obj/item/organ_holder.dm:832]: this is so much hell copypaste code, this doesn't need to be split up like this if the actual stuff being done is the same for everything here and below.

File: code/obj/augmentations.dm [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+		if(istype(chosen_organ, /obj/item/organ/brain) && !chosen_organ.robotic && !chosen_organ.synthetic)
+			return TRUE
+		else
+			return FALSE
+
+
+/datum/abilityHolder/augmentation //unused currently but laying the path for future augmentations
+	usesPoints = 0
+	regenRate = 0
+	tabName = "Body"
```
Comment (ZeWaka) [inline code/obj/augmentations.dm:202]: not a big fan of including unused code, there should be at least on example for people to copy and base their work on.

File: code/obj/augmentations.dm [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+ABSTRACT_TYPE(/obj/item/augmentation/head)
+/obj/item/augmentation/head //second abstract parent incase other augmentation types get added
+	name = "surgical augmentation parent"
+	icon_state = "augmentation"
+
+/obj/item/augmentation/head/wireless_interact //you can interact with mechanical things at range at the cost of flash vulnerability
+	name = "wireless interactor"
+	icon_state = "augmentation_wire"
+	desc = "An augmentation that allows for ranged interaction with various electronic devices."
+
```
Comment (ZeWaka) [inline code/obj/augmentations.dm:134]: shouldn't be in this file

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Comment by flrsh** [comment]
Closing as Zonespace is currently on hiatus and I don't see this getting worked on anytime soon.

**Comment by CheffieGithub** [comment]
Can I adopt this PR?

**Pull request closed without merging**